src/err/mod.rs

```diff
@@ -61,6 +61,15 @@ impl<'a> PyDowncastError<'a> {
             to: to.into(),
         }
     }
+
+    /// Compatibility API to convert the Bound variant `DowncastError` into the
+    /// gil-ref variant
+    pub(crate) fn from_downcast_err(DowncastError { from, to }: DowncastError<'a, 'a>) -> Self {
+        Self {
+            from: from.as_gil_ref(),
+            to,
+        }
+    }
```

Comment:
Good idea, I also had some similar pieces in #3706 and maybe I can build on top of this API 👍